From bdd0c4372d5a07d76b223a637308bf3a1f4c3a85 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Tue, 18 May 2010 16:09:04 +0800
Subject: [PATCH 1/2] x86: clean idle thread TS_POLLING flag

TS_POLLING will block other cpu sending reschedule ipi to
related cpu which is in idle state, and it makes rescheldule
work delayed to next timer tick, which makes our context
latency pretty terrible. So clean TS_POLLING before entering
into hyper-idle.

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 arch/x86/kernel/process.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/x86/kernel/process.c b/arch/x86/kernel/process.c
index d41a61e..f7cf2f8 100644
--- a/arch/x86/kernel/process.c
+++ b/arch/x86/kernel/process.c
@@ -479,8 +479,11 @@ static void mwait_idle(void)
 static void wrhv_idle(void)
 {
 	if (!need_resched()) {
+		current_thread_info()->status &= ~TS_POLLING;
+		smp_mb__after_clear_bit();
 		local_irq_enable();
 		vbi_idle(1);
+		current_thread_info()->status |= TS_POLLING;
 	}
 }
 #endif
-- 
1.6.5.2

